lambda_sns/lambda_function.py

```diff
@@ -0,0 +1,165 @@
+    awsAccountId = message["AWSAccountId"]
+    region = message["Region"]
+
+    #get the timestamp of the alarm/event
+    stateChangeTime = message["StateChangeTime"]
+    dt = dateutil.parser.parse(stateChangeTime)
+    timestamp = unix_time_millis(dt)
+
+     # Create the subject of the email
+    SUBJECT = "ALARM: {0} in {1}".format(alarmName, region)
+
+    # Creating the basic header of the email for TXT
+    EMAIL_BODY_TXT = """
+        You are receiving this email because your Amazon CloudWatch Alarm \"{0}\" in the \"{1}\" region has entered the ALARM state, because \"{2}\" at \"{3}\".\n
+
+        View this alarm in the AWS Management Console:\n
+        https://console.aws.amazon.com/cloudwatch/home?region=us-west-2#s=Alarms&alarm={0}\n
+
+        Alarm Details:\n
+        """.format(alarmName, region, newStateReason, timestamp )
+
+    # Creating the basic header of the email for HTML emails
+    EMAIL_BODY_HTML = """<html>
+        <head></head>
+        <body>
+        <p>You are receiving this email because your Amazon CloudWatch Alarm \"{0}\" in the \"{1}\" region has entered the ALARM state, because \"{2}\" at \"{3}\"<p/>
+
+        <p>View this alarm in the AWS Management Console:</p>
+        <p>https://console.aws.amazon.com/cloudwatch/home?region=us-west-2#s=Alarms&alarm={0}</p>
+
+        <p>Alarm Details:<br />
+        """.format(alarmName, region, newStateReason, timestamp )
+
+    # Separating out the Metrics section this will be found in the message["Trigger"]
+    METRIC_DETAILS_TXT = "\nMonitored Metric:\n"
+    METRIC_DETAILS_HTML = "<p>Monitored Metric:</p>"
+
+    # loop through the event message and create the email with the desired information
+    # Separate out the 'metric' section which is defined by 'trigger'
+
+    for key,val in message.items():
+        if "Trigger" in key:
+            trigger = message["Trigger"]
+            break
+        EMAIL_BODY_TXT += "\t{} = {}\n".format(key, val)
+        EMAIL_BODY_HTML += "<table><tr><td>{}:</td><td>{}</td></tr></table>".format(key, val)
+
+    for tkey, tval in trigger.items():
+        METRIC_DETAILS_TXT += "\t{} = {}\n".format(tkey, tval)
+        METRIC_DETAILS_HTML += "<table><tr><td>{}:</td><td>{}</td></tr></table>".format(tkey, tval)
+
+    # Now lets put the two sections together for both TXT and HTML emails.
+    EMAIL_BODY_TXT += METRIC_DETAILS_TXT
+    EMAIL_BODY_HTML += METRIC_DETAILS_HTML
+
+    send_email(EMAIL_BODY_HTML, EMAIL_BODY_TXT,SUBJECT)
+
+# simple function to just pull the 'message' from the event.  we can expand on this if needed
+def get_event_message(event):
+    messageText = event["Sns"]["Message"]
+    message = json.loads(messageText)
+    format_email(message)
+
+# The SNS will send an 'event' with a message attached.
+# Grab the event and see if it has the 'Records' in it, and if so send it off to be processed
+# If not, send an email that we have no events recorded.  Doubtful, but just a troubleshooting part.
```

Comment:
Dont understand when an event would not have records and if that is important.  Did you intent to keep this troubleshooting part in?

Reply:
My thought process is you should always plan for failures and errors as much as possible and also keep it reasonable.  This allows us to see if an event was actually sent or not.

Quick answer:  1. Yes it is important  2. Yes I intended it to be in the script.

Using this we will still get the Alert email, but can tell right away if no Event existed, or if it was a different problem.  I.E Event existed but no Alarm data might tell us there is a problem with the Alarm, while if there is no Event, it will tell us it might be a problem with SNS itself.

If no event exists, then most likely it is either 1.  AWS error  2. Someone manually firing off our Lambda function which should not happen except for known testing.  3. Something else we have not thought of yet.

So...If for any reason at all, AWS has problems with SNS(which is possible) and it does not send an event, we will still get an Alert email and know that while we got an Alert, the SNS event did not fire correctly and we can investigate that after we figure out the problem with PRD being down.